clients/src/main/java/org/apache/kafka/common/requests/VoteRequest.java

```diff
@@ -60,10 +66,60 @@ protected Struct toStruct() {
     }
 
     @Override
-    public VoteResponse getErrorResponse(int throttleTimeMs, Throwable e) {
-        VoteResponseData data = new VoteResponseData();
-        data.setErrorCode(Errors.forException(e).code());
-        return new VoteResponse(data);
+    public AbstractResponse getErrorResponse(int throttleTimeMs, Throwable e) {
+        return getErrorResponse(this.data, Errors.forException(e));
+    }
+
+    public static VoteRequestData singletonRequest(TopicPartition topicPartition,
+                                                   int candidateEpoch,
+                                                   int candidateId,
+                                                   int lastEpoch,
+                                                   long lastEpochEndOffset) {
+        return singletonRequest(topicPartition,
+            null,
+            candidateEpoch,
+            candidateId,
+            lastEpoch,
+            lastEpochEndOffset);
+    }
+
+    public static VoteRequestData singletonRequest(TopicPartition topicPartition,
+                                                   String clusterId,
+                                                   int candidateEpoch,
+                                                   int candidateId,
+                                                   int lastEpoch,
+                                                   long lastEpochEndOffset) {
+        return new VoteRequestData()
+                   .setClusterId(clusterId)
+                   .setTopics(Collections.singletonList(
+                       new VoteRequestData.VoteTopicRequest()
+                           .setTopicName(topicPartition.topic())
+                           .setPartitions(Collections.singletonList(
+                               new VoteRequestData.VotePartitionRequest()
+                                   .setPartitionIndex(topicPartition.partition())
+                                   .setCandidateEpoch(candidateEpoch)
+                                   .setCandidateId(candidateId)
+                                   .setLastOffsetEpoch(lastEpoch)
+                                   .setLastOffset(lastEpochEndOffset))
+                           )));
     }
 
+    private static VoteResponse getErrorResponse(VoteRequestData data,
```

Comment:
Maybe we can let this return `VoteResponseData` so that it can be used in `KafkaRaftClient` for error cases.